extensions/replication/queueProcessor/MultipleBackendTask.js

```diff
@@ -0,0 +1,175 @@
+                method: 'MultipleBackendTask._setupRolesOnce',
+                entry: entry.getLogInfo(),
+                roles: entryRolesString,
+            });
+            return cb(errors.BadRole);
+        }
+        this.sourceRole = entryRoles[0];
+
+        this._setupSourceClients(this.sourceRole, log);
+
+        const req = this.S3source.getBucketReplication({
+            Bucket: entry.getBucket(),
+        });
+        attachReqUids(req, log);
+        return req.send((err, data) => {
+            if (err) {
+                log.error('error getting replication configuration from S3', {
+                    method: 'MultipleBackendTask._setupRolesOnce',
+                    entry: entry.getLogInfo(),
+                    origin: 'source',
+                    peer: this.sourceConfig.s3,
+                    error: err.message,
+                    httpStatus: err.statusCode,
+                });
+                // eslint-disable-next-line no-param-reassign
+                err.origin = 'source';
+                return cb(err);
+            }
+            const replicationEnabled = data.ReplicationConfiguration.Rules
+                .some(rule => rule.Status === 'Enabled' &&
+                    entry.getObjectKey().startsWith(rule.Prefix));
+            if (!replicationEnabled) {
+                log.debug('replication disabled for object', {
+                    method: 'MultipleBackendTask._setupRolesOnce',
+                    entry: entry.getLogInfo(),
+                });
+                return cb(errors.PreconditionFailed.customizeDescription(
+                    'replication disabled for object'));
+            }
+            const roles = data.ReplicationConfiguration.Role.split(',');
+            if (roles.length !== 1) {
+                log.error('expecting a single role in bucket replication ' +
+                'configuration when replicating to an external location', {
+                    method: 'MultipleBackendTask._setupRolesOnce',
+                    entry: entry.getLogInfo(),
+                    roles,
+                });
+                return cb(errors.BadRole);
+            }
+            if (roles[0] !== entryRoles[0]) {
+                log.error('role in replication entry for source does not ' +
+                'match role in bucket replication configuration', {
+                    method: 'MultipleBackendTask._setupRolesOnce',
+                    entry: entry.getLogInfo(),
+                    entryRole: entryRoles[0],
+                    bucketRole: roles[0],
+                });
+                return cb(errors.BadRole);
+            }
+            return cb(null, roles[0]);
+        });
+    }
+
+    _getAndPutPartOnce(sourceEntry, destEntry, part, log, done) {
+        log.debug('getting object part', { entry: sourceEntry.getLogInfo() });
+        const doneOnce = jsutil.once(done);
+        // TODO handle zero-byte objects which have no parts
```

Comment:
is this still relevant?

Reply:
Yes, I will have to make the part optional, since metadata-only operations will not have parts in the locations array.